### DISCUSSION TITLE
Plotting isosurfaces for a function with a discontinuity

### DISCUSSION THREAD

**Discussion by edelopo** [pyvista/pyvista#3354] [Q&A]
Hi all! I am trying to plot some isosurfaces corresponding to a function that takes values in a circle. This means that the values actually end up going between -pi and pi, and there is a discontinuity precisely at that point. I am using the `StructuredGrid.contour` method, and its behaviour is not adequate for my case.

![image](https://user-images.githubusercontent.com/18720965/191821901-7ab686c8-f3ae-4462-a6d8-cf2fc8a9d79a.png)

As you can see in the picture, there are two parts to the plot. The smooth one corresponds to what I'm actually trying to see: the isosurface corresponding to angle 1. The blocky part is the discontinuity, where `contour` is deciding that, since there are values very close to pi, and values very close to -pi, there must be a 1 betwee. To see that this is the issue, see the plots for angles -1 and 3.1 below.

![image](https://user-images.githubusercontent.com/18720965/191822385-b5d58fde-9b45-41e2-a43d-bc11508a6d43.png)
![image](https://user-images.githubusercontent.com/18720965/191822438-2b551418-0881-46dc-a6ee-9e019dc912cb.png)

As you can see, the smooth part changes, and the blocky part stays in the plot for angle=-1, whereas the two almost coincide and there is confusion for angle=3.1.

What I would like to have to avoid this is to get some sort of control over how contour decides when to interpolate. Do you know if this is possible?

**Comment by banesullivan** [comment]
Would you be able to share this mesh file? `mesh.save('mesh.vtk')`

**Reply by edelopo** [reply]
The site doesn't seem to support .vtk files, but I can share the code I used since it is quite short:

```
import numpy as np
import pyvista as pv

a = 2
b = 3
theta = 1

def toComplex(x, y, z):
	z1 = 2*x + 2*y*1j
	z2 = 2*z + (x*x + y*y + z*z - 1)*1j
	z1 /= x*x + y*y + z*z + 1
	z2 /= x*x + y*y + z*z + 1
	return [z1, z2]

#%% Data
x, y, z = np.pi*np.mgrid[-2:2:101j, -2:2:101j, -2:2:101j]
complexPoint = toComplex(x, y, z)
argument = np.angle(complexPoint[0]**a + complexPoint[1]**b)
vol = argument
grid = pv.StructuredGrid(x, y, z)
grid["vol"] = vol.flatten()
contours = grid.contour([theta])

#%% Visualization
pv.set_plot_theme('default')
p = pv.Plotter()
p.add_mesh(contours, scalars=contours.points[:, 2], show_scalar_bar=False)
contours.save('mesh.vtk')
p.show(interactive_update=False)
```

**Reply by edelopo** [reply]
I guess I can change the extension and upload it as a .txt, and you can change it later:
[mesh.txt](https://github.com/pyvista/pyvista/files/9631406/mesh.txt)

**Comment by darikg** [comment]
It seems ill-defined to me to threshold a circular variable, since "above" and "below" don't mean anything. Maybe you could instead define an angular distance and threshold at some small value of that?

**Reply by edelopo** [reply]
I wasn't thinking of it as a threshold, just as an isosurface for a particular value. That makes a lot of sense, I think, I'm just asking when the function "points" in a specific direction. I don't fully understand your proposal, how can I threshold the contour?

**Reply by darikg** [reply]
Oh, right, sorry I misread contouring for thresholding for some reason. I think the same issue applies though. Maybe something like this?

```python
x, y, z = np.pi*np.mgrid[-2:2:101j, -2:2:101j, -2:2:101j]
complexPoint = toComplex(x, y, z)
val = complexPoint[0]**a + complexPoint[1]**b
val /= abs(val)
dot_product = np.real(val) * np.cos(theta) + np.imag(val) * np.sin(theta)  # between -1 and 1

grid = pv.StructuredGrid(x, y, z)
grid["dot_product"] = dot_product.flatten()
clipped = grid.clip_scalar(scalars='dot_product', invert=False, value=0.9)  # above 0.9 -> similar direction to theta
clipped.plot()
```

**Reply by edelopo** [reply]
This looks promising. I understand what you do with the dot product, but I don't understand what `clip_scalar` is doing. What is clipping?

**Reply by MatthewFlamm** [reply]
Interesting problem here.  Is it possible to leverage the fact that you can arbitirarily force the discontinuity in a different location?  For example, if I circularly shift the data by pi, I get a different surface that is affected by the discontinuity:

![image](https://user-images.githubusercontent.com/39341281/192036612-c1ea636e-aa4c-47a4-bb66-fdd81b820671.png)

```py
import numpy as np
import pyvista as pv

a = 2
b = 3
theta = 1

def toComplex(x, y, z):
    z1 = 2*x + 2*y*1j
    z2 = 2*z + (x*x + y*y + z*z - 1)*1j
    z1 /= x*x + y*y + z*z + 1
    z2 /= x*x + y*y + z*z + 1
    return [z1, z2]

#%% Data
x, y, z = np.pi*np.mgrid[-2:2:101j, -2:2:101j, -2:2:101j]
complexPoint = toComplex(x, y, z)
argument = np.angle(complexPoint[0]**a + complexPoint[1]**b)
vol = argument.flatten()
# shift by pi
vol_shift = vol.copy() + np.pi
# keep values between -pi and pi
vol_shift[vol_shift>np.pi] = vol_shift[vol_shift>np.pi] - 2 * np.pi

grid = pv.StructuredGrid(x, y, z)
grid["vol"] = vol
grid["vol_shift"] = vol_shift
contours = grid.contour([theta], scalars='vol')
# account for pi shift in contour
contours_shift = grid.contour([theta-np.pi], scalars='vol_shift')
#%% Visualization
pv.set_plot_theme('default')
p = pv.Plotter()
p.add_mesh(contours, style='wireframe', color='red')
p.add_mesh(contours_shift, style='wireframe', color='blue')
contours.save('mesh.vtk')
p.show(interactive_update=False)
```

**Reply by edelopo** [reply]
This is an interesting idea, but not ideal, since you still get "two" surfaces and the one corresponding to the branch cut is still clunky :(
